src/base/InstanceId.h

```diff
@@ -39,33 +39,43 @@ class InstanceId
     const char * prefix() const;
 
 public:
-    Value value = 0; ///< instance identifier
+    Value value = Value(); ///< instance identifier
 
 private:
     InstanceId(const InstanceId &); ///< not implemented; IDs are unique
     InstanceId& operator=(const InstanceId &); ///< not implemented
 };
 
-/// convenience macro to instantiate Class-specific stuff in .cc files
-#define InstanceIdDefinitions(Class, pfx) \
+/// 2-parameter instantiation macro (for private use only).
+/// In other sources use InstanceIdDefinitions() instead.
+#define InstanceIdDefinitions3(Class, ValueType, pfx) \
     template<> const char * \
-    InstanceId<Class>::prefix() const { \
+    InstanceId<Class, ValueType>::prefix() const { \
         return pfx; \
     } \
     template<> std::ostream & \
-    InstanceId<Class>::print(std::ostream &os) const { \
+    InstanceId<Class, ValueType>::print(std::ostream &os) const { \
         return os << pfx << value; \
     } \
     template<> void \
-    InstanceId<Class>::change() { \
-        static InstanceId<Class>::Value Last = 0; \
+    InstanceId<Class, ValueType>::change() { \
+        static auto Last = Value(); \
         value = ++Last ? Last : ++Last; \
     }
 
+/// 1-parameter instantiation macro (for private use only).
+/// In other sources use InstanceIdDefinitions() instead.
+#define InstanceIdDefinitions2(Class, pfx) InstanceIdDefinitions3(Class, InstanceIdDefaultValueType, pfx)
+
+#define GetInstanceIdMacro(_1, _2, _3, MacroName, ...) MacroName
+
+/// convenience macro to instantiate Class-specific stuff in .cc files
+#define InstanceIdDefinitions(...) GetInstanceIdMacro(__VA_ARGS__, InstanceIdDefinitions3, InstanceIdDefinitions2)(__VA_ARGS__)
```

Comment:
Does reordering InstanceIdDefinitions3() and adding "..." allow us to avoid the GetInstanceIdMacro() trick?

```suggestion
#define InstanceIdDefinitions3(Class, pfx, ValueType, ...) \
    ... existing InstanceIdDefinitions3 body here ...
#define InstanceIdDefinitions(...) InstanceIdDefinitions3(__VA_ARGS__, InstanceIdDefaultValueType)
```

If something like this allows us to avoid the GetInstanceIdMacro() trick, then I think we should avoid it.

Reply:
Furthermore, can we simply use `<__VA_ARGS__>` instead of `<Class, ValueType>` in InstanceIdDefinitions3()? If yes, then we can remove InstanceIdDefaultValueType declaration:

```
#define AnyInstanceIdDefinitions(pfx, ...) \
 ... existing InstanceIdDefinitions3 body here but using <__VA_ARGS__> ...
```

InstanceIdDefinitions3() should be renamed to something like AnyInstanceIdDefinitions() in that case.

However, I am not sure the removal of InstanceIdDefaultValueType is worth spreading `__VA_ARGS__` like that. Your call.